### PR TITLE
MOTECH-2346: Fixed: It's not possible to save Commcare settings after change Configuration Name

### PR DESCRIPTION
**Pull request by pkornowski**
-  Commcare configurations on the server are identified by name. If we changed name of Configuration and click "Save" button, then on the server we searching configuration identified by new name, which it wasn't yet. I added new parameter "oldName" which stores name before change. When we update our configuration, then on the server side, configuration is searching by not updated name.
-  When we change "Event Strategy" of commcare settings we get error in console about $scope.saveSettings is not exists. It was because of "Event strategy" parameter had autosave, which invoked saveSettings function. It's not neccessary, because this parameter is saving like other parameter during clicks "Save" button, so I deleted this autosave.

### CODE REVIEW

**Comment by kmadej** [comment]
Do not add field 'oldName' to Config class, just send this name as parameter

Additionally please add a unit test.

File: commcare/src/main/java/org/motechproject/commcare/config/Configs.java [hunk 1]
```diff
@@ -132,8 +132,8 @@ public boolean hasDefault() {
      *
      * @param config  the update source
      */
-    public void updateConfig(Config config) {
-        deleteConfig(config.getName());
+    public void updateConfig(Config config, String oldName) {
```
Comment (kmadej) [inline commcare/src/main/java/org/motechproject/commcare/config/Configs.java:136]: add javadoc refering new parameter

File: commcare/src/main/java/org/motechproject/commcare/web/ConfigController.java [hunk 2]
```diff
@@ -53,8 +54,8 @@ public void makeDefault(@RequestBody String name) {
     @ResponseStatus(HttpStatus.OK)
     @RequestMapping(method = RequestMethod.POST)
     @ResponseBody
-    public Config saveConfig(@RequestBody Config config) throws CommcareConnectionFailureException {
-        return configService.saveConfig(config);
+    public Config saveConfig(@RequestBody Config config, @RequestParam("oldName") String oldName) throws CommcareConnectionFailureException {
```
Comment (kmadej) [inline commcare/src/main/java/org/motechproject/commcare/web/ConfigController.java]: this parameter in RequestParam annotation is redundant

File: commcare/src/test/java/org/motechproject/commcare/it/CommcareTasksIntegrationBundleIT.java [hunk 3]
```diff
@@ -313,8 +314,8 @@ private HttpResponse sendMockForm() throws IOException, InterruptedException {
         return getHttpClient().execute(httpPost);
     }
 
-    private HttpResponse createConfiguration(Config config) throws IOException, InterruptedException {
-        HttpPost httpPost = new HttpPost(String.format("http://localhost:%d/commcare/configs", PORT));
+    private HttpResponse createConfiguration(Config config, String oldName) throws IOException, InterruptedException {
+        HttpPost httpPost = new HttpPost(String.format("http://localhost:%d/commcare/configs?oldName=%s", PORT, oldName));
```
Comment (kmadej) [inline commcare/src/test/java/org/motechproject/commcare/it/CommcareTasksIntegrationBundleIT.java]: why you used this request to create config ? I thought that there are two requests, one for create and second for update config.

File: commcare/src/test/java/org/motechproject/commcare/web/ConfigControllerTest.java [hunk 4]
```diff
@@ -79,10 +79,15 @@ public void shouldSetDefault() {
     public void testSaveSettings() throws BundleException, CommcareAuthenticationException, CommcareConnectionFailureException {
 
         Config config = ConfigsUtils.prepareConfigThree();
+        String oldName = config.getName();
+        controller.saveConfig(config, oldName);
 
-        controller.saveConfig(config);
+        verify(configService).saveConfig(config, oldName);
+
+        config.setName("updatedName");
```
Comment (kmadej) [inline commcare/src/test/java/org/motechproject/commcare/web/ConfigControllerTest.java]: apart from fact that this test is rather redundant, I do not understand your changes. Why after veryfing you are setting name in config and then again use verify ? It doesn't make sense

**Comment by kmadej** [comment]
Additionally, please add an unit test to CommcareConfigServiceImplTest.

**Comment by pkornowski** [comment]
I split saveConfig. If we adding new configuration, then on the server is sending only config now. If we updating configuration, then we sending config and old name of configuration. I added test for save updated config and for save new config in CommcareConfigServiceImplTest.

File: commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java [hunk 5]
```diff
@@ -10,12 +10,21 @@
 public interface CommcareConfigService {
 
     /**
+     * Updates the given configuration.
+     *
+     * @param configuration  the configuration to be saved
+     * @param oldName   name of configuration before update
```
Comment (pgesek) [inline commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java]: the update

File: commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java [hunk 6]
```diff
@@ -10,12 +10,21 @@
 public interface CommcareConfigService {
 
     /**
+     * Updates the given configuration.
```
Comment (pgesek) [inline commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java]: Used for changing configuration names

File: commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java [hunk 7]
```diff
@@ -10,12 +10,21 @@
...
+     *
+     * @param configuration  the configuration to be saved
+     * @param oldName   name of configuration before update
+     * @return  the saved configuration
+     */
+    Config saveUpdatedConfig(Config configuration, String oldName) throws CommcareConnectionFailureException;
```
Comment (pgesek) [inline commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java]: rename this method to updateConfig

File: commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java [hunk 8]
```diff
@@ -10,12 +10,21 @@
...
+
+    /**
      * Creates the given configuration.
      *
      * @param configuration  the configuration to be saved
      * @return  the saved configuration
      */
-    Config saveConfig(Config configuration) throws CommcareConnectionFailureException;
+    Config saveNewConfig(Config configuration) throws CommcareConnectionFailureException;
```
Comment (pgesek) [inline commcare/src/main/java/org/motechproject/commcare/service/CommcareConfigService.java]: rename this method to addConfig

File: commcare/src/main/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImpl.java [hunk 9]
```diff
@@ -78,13 +78,42 @@ public CommcareConfigServiceImpl(@Qualifier("commcareAPISettings") SettingsFacad
     }
 
     @Override
-    public Config saveConfig(Config config) throws CommcareConnectionFailureException {
+    public Config saveUpdatedConfig(Config config, String oldName) throws CommcareConnectionFailureException {
```
Comment (pgesek) [inline commcare/src/main/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImpl.java]: there is too much duplication between the methods - I might be wrong, but isn't saveNewConfig(config) the same as saveUpdatedConfig(config, config.getName()) ?

File: commcare/src/main/resources/webapp/js/controllers.js [hunk 10]
```diff
@@ -247,6 +249,7 @@
             copy.forwardSchema = config.forwardSchema;
             copy.forwardStubs = config.forwardStubs;
             copy.forwardCases = config.forwardCases;
+            $scope.oldName = copy.name;
```
Comment (pgesek) [inline commcare/src/main/resources/webapp/js/controllers.js]: don't like the idea of storing this in scope, while everything else is returned by the method. Can't we return an object of {copy, oldName} instead?

File: commcare/src/main/resources/webapp/js/controllers.js [hunk 11]
```diff
@@ -506,9 +510,36 @@
             );
         };
 
-        $scope.saveConfig = function(element) {
+        $scope.saveUpdatedConfig = function(element) {
```
Comment (pgesek) [inline commcare/src/main/resources/webapp/js/controllers.js]: updateConfig

File: commcare/src/main/resources/webapp/js/controllers.js [hunk 12]
```diff
@@ -529,6 +560,15 @@
                 });
         };
 
+        $scope.saveConfig = function(element) {
+            blockUI();
+            if($scope.oldName!=="") {
```
Comment (pgesek) [inline commcare/src/main/resources/webapp/js/controllers.js]: just change to if ($scope.oldName)

File: commcare/src/test/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImplTest.java [hunk 13]
```diff
@@ -63,4 +71,35 @@ public void shouldReturnBlankUrlForNullOrBlankServerUrl() {
...
+
+    @Test
+    public void shouldSaveNewConfiguration() throws CommcareConnectionFailureException {
+
+        Config config = configService.create();
+        config.setName("newName");
+
+        Config savedConfig = configService.saveNewConfig(config);
+        assertEquals(config.getName(), savedConfig.getName());
+    }
```
Comment (pgesek) [inline commcare/src/test/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImplTest.java:81]: new line after this

File: commcare/src/test/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImplTest.java [hunk 14]
```diff
@@ -6,19 +6,27 @@
...
+import java.util.ArrayList;
+import java.util.List;
 
 import static org.junit.Assert.assertEquals;
 import static org.mockito.Mockito.when;
 
 @RunWith(MockitoJUnitRunner.class)
 public class CommcareConfigServiceImplTest {
 
+
```
Comment (pgesek) [inline commcare/src/test/java/org/motechproject/commcare/service/impl/CommcareConfigServiceImplTest.java]: no need for these empty lines

**Pull request merged**